iroh-memstore/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+//! In-memory store for iroh.
+//!
+//! This is an in-memory store for iroh, it implements the internal RPC interface for a
+//! store and thus can be used as a drop-in replacement for the store.  It is not optimised
+//! at all.
+//!
+//! This store has no concurrency at all, all RPC requests are handled sequentially.
+
+use std::sync::{Arc, Mutex};
+
+use ahash::AHashMap;
+use anyhow::Result;
+use bytes::Bytes;
+use cid::Cid;
+use futures::stream::Stream;
+use iroh_rpc_client::{ChannelTypes, ServerSocket, StoreServer, HEALTH_POLL_WAIT};
+use iroh_rpc_types::store::{
+    GetLinksRequest, GetLinksResponse, GetRequest, GetResponse, GetSizeRequest, GetSizeResponse,
+    HasRequest, HasResponse, PutManyRequest, PutRequest, StoreAddr, StoreRequest, StoreService,
+};
+use iroh_rpc_types::{VersionRequest, VersionResponse, WatchRequest, WatchResponse};
+use quic_rpc::server::RpcServerError;
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+use tracing::{debug, error, info};
+
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Debug)]
+pub struct MemStoreHandle {
+    shutdown_tx: oneshot::Sender<()>,
+    handle: JoinHandle<()>,
+}
+
+impl MemStoreHandle {
+    pub fn shutdown(self) -> JoinHandle<()> {
+        // Failing to send the shutdown signal means the task already died somehow.  We can
+        // still return the JoinHandle which should complete when awaited.
+        self.shutdown_tx.send(()).ok();
+        self.handle
+    }
+}
+
+pub async fn spawn(addr: StoreAddr) -> Result<MemStoreHandle> {
+    let server = iroh_rpc_client::create_server::<StoreService>(addr).await?;
+    let (shutdown_tx, mut shutdown_rx) = oneshot::channel();
+    let mut store = MemStore::new();
+    let handle = tokio::spawn(async move {
+        loop {
+            tokio::select! {
+                biased;
+                _ = &mut shutdown_rx => {
+                    info!("Shutting down MemStore");
+                    break
+                },
+                req = server.accept_one() => store.dispatch_request(&server, req).await,
+            };
+        }
+        // TODO: The quick-rpc RpcServer needs to be improved so it can be shut down without
+        // dropping and we can await the completion.
+    });
+    Ok(MemStoreHandle {
+        shutdown_tx,
+        handle,
+    })
+}
+
+#[derive(Debug, Default, Clone)]
+struct MemStore {
+    inner: Arc<Mutex<InnerMemStore>>,
+}
+
+#[derive(Debug, Default)]
+struct InnerMemStore {
+    data: AHashMap<Cid, Bytes>,
```

Comment:
So no id/cid mapping? My hunch is that this is worth it even for the mem store, but it can wait for later.

Reply:
In the normal store we use u64 ids all over the place. Cids are only used in one place, to look up the corresponding id. The reason for that is that cids are quite large and unwieldy, whereas u64s are, well u64s.

For sure this is not needed in the mem store at this time, but eventually it will be. Cids in memory are something like 100 bytes.

Reply:
I can only see a space saving if links can also be an id instead of full Cid.  But that means all links must exist in the store and seems a bit unlikely.  Or am I still getting this wrong?